src/Catalog/Icons/IconProcessor.cs

```diff
@@ -0,0 +1,274 @@
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<Uri> CopyIconFromExternalSource(
+            Stream iconDataStream,
+            IStorage destinationStorage,
+            string destinationStoragePath,
+            CancellationToken cancellationToken,
+            string packageId,
+            string normalizedPackageVersion)
+        {
+            var destinationUri = destinationStorage.ResolveUri(destinationStoragePath);
+
+            var iconData = await GetStreamBytesAsync(iconDataStream, MaxExternalIconSize, cancellationToken);
+            if (iconData == null)
+            {
+                return null;
+            }
+
+            var contentType = DetermineContentType(iconData, onlyGallerySupported: false);
+            if (string.IsNullOrWhiteSpace(contentType))
+            {
+                _logger.LogInformation("Failed to determine image type.");
+                return null;
+            }
+            _logger.LogInformation("Content type for {PackageId} {PackageVersion} {ContentType}", packageId, normalizedPackageVersion, contentType);
+            var content = new ByteArrayStorageContent(iconData, contentType, DefaultCacheControl);
+            await destinationStorage.SaveAsync(destinationUri, content, cancellationToken);
+            _telemetryService.TrackExternalIconIngestionSuccess(packageId, normalizedPackageVersion);
+            return destinationUri;
+        }
+
+        public async Task DeleteIcon(
+            Storage destinationStorage,
+            string destinationStoragePath,
+            CancellationToken cancellationToken,
+            string packageId,
+            string normalizedPackageVersion)
+        {
+            _logger.LogInformation("Deleting icon blob {IconPath}", destinationStoragePath);
+            if (destinationStorage.Exists(destinationStoragePath))
+            {
+                var iconUri = new Uri(destinationStorage.BaseAddress, destinationStoragePath);
+                await destinationStorage.DeleteAsync(iconUri, cancellationToken);
+            }
+        }
+
+        public async Task<Uri> CopyEmbeddedIconFromPackage(
+            Stream packageStream,
+            string iconFilename,
+            IStorage destinationStorage,
+            string destinationStoragePath,
+            CancellationToken cancellationToken,
+            string packageId,
+            string normalizedPackageVersion)
+        {
+            var iconPath = PathUtility.StripLeadingDirectorySeparators(iconFilename);
+            var destinationUri = destinationStorage.ResolveUri(destinationStoragePath);
+
+            await ExtractAndStoreIconAsync(packageStream, iconPath, destinationStorage, destinationUri, cancellationToken, packageId, normalizedPackageVersion);
+            return destinationUri;
+        }
+
+        private async Task ExtractAndStoreIconAsync(
```

Comment:
>ExtractAndStoreIconAsync [](start = 27, length = 24)

The fact that this does not share code with Catalog2Dnx is concerning. Couldn't the implementations diverge easily?

Reply:
Added issue to track converging those: https://github.com/NuGet/Engineering/issues/2748